File: RB101_RB109_small_problems/medium_1/rotation_p3.rb
=begin

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: integer
-  Output: maximum rotation of the integer argument

---

**Problem Domain:**

Maximum rotation of a number is the result of iteratively rotating the number:
- 1st rotation: rotate entire number
- 2nd rotation: rotate entire number, except for the 1st digit
- 3rd rotation: rotate entire number, except for the first 2 digits
- 4th rotation: rotate entire number, except for the first 3 digits
- Continue rotations in this pattern until no more rotations can be performed.

---

**Implicit Requirements:**

- Can (and should) use the #rotate_rightmost_digits method from the previous exercise.
- Leading zeros should be dropped from the result
- Do not worry about handling multiple zeros.

---

**Clarifying Questions:**

1. Will the input integer argument always be positive?
  - Yes, assume integer argument will always be positive.

---

**Mental Model:**

- Determine number length (how many digits)
- Perform [number length] number of iterations (i = 0..number length - 1)
  - Rotate the number using the rotate_rightmost_digits.
    - # of rightmost digits to rotate will be the number length minus the current iteration
- Return fully rotated number.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

max_rotation(735291) == 321579
max_rotation(3) == 3
max_rotation(35) == 53
max_rotation(105) == 15 # the leading zero gets dropped
max_rotation(8_703_529_146) == 7_321_609_845

---

Data Structure
--------------

- Use integers for looping and iterations. Previously created methods will use arrays and strings for rotation.

---

Algorithm
---------

- Determine number length (how many digits in input integer)
- Perform [number length] number of iterations (i = 0...number length)
  - Rotate the number using the rotate_rightmost_digits.
    - # of rightmost digits to rotate will be the number length minus the current iteration
  - Reassign number to output of most recent rotation.
- Return fully rotated number.

Code
----

=end

def rotate_array(input_arr)
  return [] if input_arr.empty?
  rotated_array = input_arr.clone
  rotated_array << rotated_array.shift
end

def rotate_rightmost_digits(number, n)
  all_digits = number.to_s.chars
  all_digits[-n..-1] = rotate_array(all_digits[-n..-1])
  all_digits.join.to_i
end

def max_rotation(int)
  number_of_digits = int.to_s.length

  number_of_digits.times do |i|
    current_rotation_index = number_of_digits - i
    int = rotate_rightmost_digits(int, current_rotation_index)
  end

  int
end

p max_rotation(735291) == 321579
p max_rotation(3) == 3
p max_rotation(35) == 53
p max_rotation(105) == 15 # the leading zero gets dropped
p max_rotation(8_703_529_146) == 7_321_609_845

#Further Exploration
=begin

The 3 part approach did make the problem easier to understand, but my mind first thought of a different way to complete the problem that wouldn't use the #rotate_rightmost_digits method.

If we did not have the #rotate_rightmost_digits or #rotate_array methods, my approach and solution to this problem would be different.

My approach (assuming we could not use Array#rotate) would have been to split the integer into an array of it's digits. 
Then loop until the digits array is empty. 
Each loop would move the first element to the end of the array, destructively shift the new first element to a different array, then rinse & repeat until the digits array is empty.
Then the new array of rotated digits would be joined together and converted to an integer.

I wrote this method below. This method above would also preserve consecutive zeros.

=end

def max_rotation(int)
  digits = int.digits.reverse
  rotated_digits = []

  until digits.empty?
    digits << digits.shift
    rotated_digits << digits.shift
  end

  rotated_digits.join.to_i
end

p max_rotation(735291) == 321579
p max_rotation(3) == 3
p max_rotation(35) == 53
p max_rotation(105) == 15 # the leading zero gets dropped
p max_rotation(8_703_529_146) == 7_321_609_845
p max_rotation(1000100100101010010)

File: JS210_small_problems/medium_1/rotation_p3.js
"use strict";
/*

Take the number 735291 and rotate it by one digit to the left, getting 352917. Next, keep the first digit fixed in place and rotate the remaining digits to get 329175. Keep the first two digits fixed in place and rotate again to get 321759. Keep the first three digits fixed in place and rotate again to get 321597. Finally, keep the first four digits fixed in place and rotate the final two digits to get 321579. The resulting number is called the maximum rotation of the original number.

Write a function that takes an integer as an argument and returns the maximum rotation of that integer. You can (and probably should) use the rotateRightmostDigits function from the previous exercise.

PEDAC Template
==============

(Understand the) Problem
------------------------

-  Inputs: a single integer
-  Output: return a number that represents the maximum rotation of the input integer

Note: we can uset the rotateRightmostDigits function from the last exercise

---

**Problem Domain:**

The maximum rotation of an integer is performed by:
- iterating from the number of digits in the integer down to 1. call the number `digitsToRotate`
  - on each iteration, rotating the right most `digitsToRotate` digits in the integer.
  - reassigning the integer to the result of the above rotation

---

**Explicit Requirements:**
- The argument should be an integer
- The maximum rotation of the integer should be performed and returned as a number
- We can use the rotateRightmostDigits function from the previous exercise

---


**Implicit Requirements:**
 - the function should be named maxRotation
 - any leading zeros in the final returned number should be dropped
 - any leading zeros that occur in the interim rotations should NOT be dropped.
---

**Clarifying Questions:**

1. Can we assume the input is always a non-negative integer?
  Yes.
2. To confirm, zeros should be maintained in the returned number as long as they don't become leading zeros in the final returned number, right?
  Yes.

---

**Mental Model:**

- Split the input into an array of digits (as strings).
- Perform the max rotation on the array of digits.
- Rejoin the digits into a string, convert to a number, and return the result.

---

Examples / Test Cases / Edge Cases
----------------------------------

**Examples:**

maxRotation(735291);          // 321579
maxRotation(3);               // 3
maxRotation(35);              // 53
maxRotation(105);             // 15 -- the leading zero gets dropped
maxRotation(8703529146);      // 7321609845
maxRotation(1430);            // 1430 -> 4301 -> 4013 -> 4031
maxRotation(10004);           // 10004 -> 00041 -> 00410 -> 00104 -> 00140 -> 140
maxRotation(10043);           // 10043 -> 00431 -> 04310 -> 04103 -> 04130 -> 4130

---

Data Structure
--------------

- Use strings to represent each digit. This will allow us to maintain zeros (as long as they don't become leading zeros) in the result.
- Use an array to contain the strings of digits. This will allow use of list processing methods.

---

Algorithm
---------

- Split the input into an array of digits (as strings).
- Use an empty array to contain the max rotation digits. We'll only push elements to this array once their position is set in the overall "number".
- Perform the max rotation on the array of digits.
  - Iterate from 1 to the length of the array.
    - Rotate the entire digits array to the left one spot.
    - Shift the first element from the digits array onto the end of the max rotation result array.
- Join the max rotation result array into a string, convert to a number, and return the result.

Code
----
*/

function maxRotation(integer) {
  let maxRotationArr = [];
  let digits = String(integer).split('');
  
  for (let iteration = digits.length; iteration > 0; iteration -= 1) {
    digits = digits.concat(digits.splice(0, 1));
    maxRotationArr.push(digits.shift());
  }

  return Number(maxRotationArr.join(''));
}

maxRotation(735291);          // 321579
maxRotation(3);               // 3
maxRotation(35);              // 53
maxRotation(105);             // 15 -- the leading zero gets dropped
maxRotation(8703529146);      // 7321609845
maxRotation(1430);            // 1430 -> 4301 -> 4013 -> 4031
maxRotation(10004);           // 10004 -> 00041 -> 00410 -> 00104 -> 00140 -> 140
maxRotation(10043);           // 10043 -> 00431 -> 04310 -> 04103 -> 04130 -> 4130